cub/cub/device/device_transform.cuh

```diff
@@ -0,0 +1,108 @@
+  /// input elements and writing the result to the corresponding output element. No guarantee is given on the identity
+  /// (i.e. address) of the objects passed to the call operator of the transformation operation.
+  ///
+  /// @param count The number of elements in each input stream.
+  /// @param inputs A tuple of iterators to the input streams where count elements are read from each. The iterators'
+  /// value types must be trivially relocatable.
+  /// @param output An iterator to the output stream where count results are written to.
+  /// @param transform_op An n-ary function object, where n is the number of input streams. The input iterators' value
+  /// types must be convertible to the parameters of the function object's call operator. The return type of the call
+  /// operator must be assignable to the dereferenced output iterator.
+  template <typename... RandomAccessIteratorsIn, typename RandomAccessIteratorOut, typename TransformOp>
+  CUB_RUNTIME_FUNCTION static cudaError_t ManyToOne(
+    int count,
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> inputs,
+    RandomAccessIteratorOut output,
+    TransformOp transform_op,
+    cudaStream_t stream = nullptr)
+  {
+    CUB_DETAIL_NVTX_RANGE_SCOPE("cub::DeviceTransform::TransformMany");
+    return detail::transform::
+      dispatch_t<true, int, ::cuda::std::tuple<RandomAccessIteratorsIn...>, RandomAccessIteratorOut, TransformOp>::
+        dispatch(count, std::move(inputs), output, std::move(transform_op), stream);
+  }
+
+  template <typename... RandomAccessIteratorsIn, typename RandomAccessIteratorOut, typename TransformOp>
+  CUB_RUNTIME_FUNCTION static cudaError_t ManyToOne(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    int count,
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> inputs,
+    RandomAccessIteratorOut output,
+    TransformOp transform_op,
+    cudaStream_t stream = nullptr)
+  {
+    if (d_temp_storage == nullptr)
+    {
+      temp_storage_bytes = 1;
+      return cudaSuccess;
+    }
+
+    return ManyToOne(count, std::move(inputs), output, std::move(transform_op), stream);
+  }
+
+  /// Like \ref ManyToOne, but the objects passed to the call operator of the transformation operation are guaranteed to
+  /// reside in the input streams and are never copied.
+  template <typename... RandomAccessIteratorsIn, typename RandomAccessIteratorOut, typename TransformOp>
+  CUB_RUNTIME_FUNCTION static cudaError_t ManyToOneStableParameterAddresses(
+    int count,
+    ::cuda::std::tuple<RandomAccessIteratorsIn...> inputs,
+    RandomAccessIteratorOut output,
+    TransformOp transform_op,
+    cudaStream_t stream = nullptr)
+  {
+    CUB_DETAIL_NVTX_RANGE_SCOPE("cub::DeviceTransform::TransformMany");
+    return detail::transform::
+      dispatch_t<false, int, ::cuda::std::tuple<RandomAccessIteratorsIn...>, RandomAccessIteratorOut, TransformOp>::
+        dispatch(count, std::move(inputs), output, std::move(transform_op), stream);
```

Comment:
question: most of the transform dispatch code copy iterators and transform op by value, why moving them here?

Reply:
Should also be `cuda::std::move`

Generally, we at some point want to support move only iterators, so that might be the right way going forward.

Reply:
My intention was to trigger exactly this discussion. Historically, iterators and function objects were supposed to be cheap value types, but I can imagine more complex cases. E.g., a `transform_iterator` with an extended lambda or an `nvstd::function` inside. The `move` does not hurt but it could improve the compile- and runtime if we didn't pull a copy on each function call.

If it's fine for you, I would add a move on all layers this PR introduces.

Reply:
Due to lack of answers, I am using `cuda::std::move` in all places now.